audio/out/ao_aaudio.c

```diff
@@ -0,0 +1,454 @@
+                ao->format = AF_FORMAT_S32;
+                format = AAUDIO_FORMAT_PCM_I32;
+            } else {
+                ao->format = AF_FORMAT_S16;
+                format = AAUDIO_FORMAT_PCM_I16;
+            }
+        } else {
+            ao->format = AF_FORMAT_S16;
+            format = AAUDIO_FORMAT_PCM_I16;
+        }
+
+        if (p->device_api >= 32) {
+            struct mp_chmap_sel sel = {0, .tmp = p};
+            aaudio_channel_mask_t channel_mask = AAUDIO_CHANNEL_INVALID;
+
+            for (int i = 1; i < MP_ARRAY_SIZE(aaudio_chmaps); i++)
+                mp_chmap_sel_add_map(&sel, &aaudio_chmaps[i]);
+
+            if (!ao_chmap_sel_adjust(ao, &sel, &ao->channels)) {
+                MP_ERR(ao, "Failed to find channel map");
+                goto err;
+            }
+
+            for (uint8_t i = 0; i < MP_ARRAY_SIZE(aaudio_chmaps); i++) {
+                if (mp_chmap_equals(&aaudio_chmaps[i], &ao->channels)) {
+                    channel_mask = aaudio_masks[i];
+                    break;
+                }
+            }
+            assert(channel_mask != AAUDIO_CHANNEL_INVALID);
+
+            p->AAudioStreamBuilder_setChannelMask(builder, channel_mask);
+        } else {
+            p->AAudioStreamBuilder_setChannelCount(builder, ao->channels.num);
+        }
+
+        p->AAudioStreamBuilder_setDeviceId(builder, p->device_id);
+        p->AAudioStreamBuilder_setDirection(builder, AAUDIO_DIRECTION_OUTPUT);
+        p->AAudioStreamBuilder_setSharingMode(builder,
+                                              (ao->init_flags & AO_INIT_EXCLUSIVE) ?
+                                                  AAUDIO_SHARING_MODE_EXCLUSIVE :
+                                                  AAUDIO_SHARING_MODE_SHARED);
+        p->AAudioStreamBuilder_setFormat(builder, format);
+        p->AAudioStreamBuilder_setSampleRate(builder, ao->samplerate);
+        p->AAudioStreamBuilder_setErrorCallback(builder, error_callback, ao);
+        p->AAudioStreamBuilder_setBufferCapacityInFrames(builder, p->buffer_capacity);
+        p->AAudioStreamBuilder_setPerformanceMode(builder, p->performance_mode);
+        p->AAudioStreamBuilder_setDataCallback(builder, data_callback, ao);
+
+        if (p->device_api >= 28) {
+            p->AAudioStreamBuilder_setContentType(builder, (ao->init_flags &
+                                                            AO_INIT_MEDIA_ROLE_MUSIC) ?
+                                                               AAUDIO_CONTENT_TYPE_MUSIC :
+                                                               AAUDIO_CONTENT_TYPE_MOVIE);
+            p->AAudioStreamBuilder_setUsage(builder, AAUDIO_USAGE_MEDIA);
+        }
+
+        p->builder = builder;
+
+        if ((result = p->AAudioStreamBuilder_openStream(p->builder, &p->stream)) >= 0) {
+            if (p->device_api < 32) {
+                int32_t channel_count = p->AAudioStream_getChannelCount(p->stream);
```

Comment:
why do this here? can the channel count change after opening the stream?

Reply:
Actually for older devices we need to select the chmap after opening the stream since `AAudioStreamBuilder_setChannelCount()` just drops the channels that are above the device's supported channel count.
I've removed the code calling `AAudioStreamBuilder_setChannelCount()` above